pkg/node/connection_pool.go

```diff
@@ -4,43 +4,115 @@
 package node
 
 import (
+	"context"
 	"sync"
+
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/provider"
+	"storj.io/storj/pkg/transport"
 )
 
+// PoolError is the class of errors for the connection pool
+var PoolError = errs.Class("pool error")
+
 // ConnectionPool is the in memory implementation of a connection Pool
 type ConnectionPool struct {
+	tc    transport.Client
 	mu    sync.RWMutex
-	cache map[string]interface{}
+	items map[string]*Conn
 }
 
+// Conn is the connection that is stored in the connection pool
+type Conn struct {
+	Client pb.NodesClient
+	grpc   *grpc.ClientConn
+	addr   string
+	dial   sync.Once
+	err    error
+}
+
+// NewConn intitalizes a new Conn struct with the provided address, but does not iniate a connection
+func NewConn(addr string) *Conn { return &Conn{addr: addr} }
+
 // NewConnectionPool initializes a new in memory pool
-func NewConnectionPool() *ConnectionPool {
+func NewConnectionPool(identity *provider.FullIdentity) *ConnectionPool {
 	return &ConnectionPool{
-		cache: make(map[string]interface{}),
+		tc:    transport.NewClient(identity),
+		items: make(map[string]*Conn),
 		mu:    sync.RWMutex{},
 	}
 }
 
 // Add takes a node ID as the key and a node client as the value to store
-func (pool *ConnectionPool) Add(key string, value interface{}) error {
-	pool.mu.Lock()
-	defer pool.mu.Unlock()
-	pool.cache[key] = value
+func (p *ConnectionPool) Add(key string, value interface{}) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	c, ok := value.(*Conn)
+	if !ok {
+		return PoolError.New("invalid value for connection pool")
+	}
+
+	p.items[key] = c
+
 	return nil
 }
 
 // Get retrieves a node connection with the provided nodeID
 // nil is returned if the NodeID is not in the connection pool
-func (pool *ConnectionPool) Get(key string) (interface{}, error) {
-	pool.mu.Lock()
-	defer pool.mu.Unlock()
-	return pool.cache[key], nil
+func (p *ConnectionPool) Get(key string) (interface{}, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	i, ok := p.items[key]
+	if !ok {
+		return nil, nil
+	}
+
+	return i, nil
 }
 
 // Remove deletes a connection associated with the provided NodeID
-func (pool *ConnectionPool) Remove(key string) error {
-	pool.mu.Lock()
-	defer pool.mu.Unlock()
-	pool.cache[key] = nil
-	return nil
+func (p *ConnectionPool) Remove(key string) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	i, ok := p.items[key]
+	if !ok {
+		return nil
+	}
+
+	delete(p.items, key)
+
+	return i.grpc.Close()
```

Comment:
>Close tears down the ClientConn and all underlying connections. 

It would close the connection for the other go routine. I think this is acceptable. Thoughts ?

Reply:
Would it be easiest to do some sort of inactivity timeout and remove the manual Remove() method?

Reply:
We will want to disconnect from nodes that we don't plan on contacting anymore i.e. it has failed Ping too many times